### PR TITLE
issue #11571 gfx_draw_string_left

### PR DESCRIPTION
**Pull request by modfort**
i have overload the fonction gfx_draw_string_left by replacing int x , int y with const ScreenCoordsXY& in the File Text.cpp:
void gfx_draw_string_left(rct_drawpixelinfo* dpi, rct_string_id format, void* args, uint8_t colour, const ScreenCoordsXY& coords)
after that i had to declare the new function in Dranwing.h and test it in interface/Widget.cpp

### CODE REVIEW

File: src/openrct2/drawing/Drawing.h [hunk 1]
```diff
@@ -13,6 +13,7 @@
 #include "../common.h"
 #include "../interface/Colour.h"
 #include "../interface/ZoomLevel.hpp"
+#include "../world/Location.hpp"
```
Comment (duncanspumpkin) [inline src/openrct2/drawing/Drawing.h:16]: You don't need to include this header if you forward declare the screencoordsxy.

`struct ScreenCoordsXY;`

File: src/openrct2/drawing/Drawing.h [hunk 2]
```diff
@@ -14,7 +14,9 @@
 #include "../interface/Colour.h"
 #include "../interface/ZoomLevel.hpp"
 
+
```
Comment (tupaschoal) [inline src/openrct2/drawing/Drawing.h:17]: Please revert his addition

File: src/openrct2/drawing/Text.cpp [hunk 3]
```diff
@@ -155,6 +155,10 @@ void gfx_draw_string_left(rct_drawpixelinfo* dpi, rct_string_id format, void* ar
 {
     DrawTextCompat(dpi, x, y, format, args, colour, TextAlignment::LEFT);
 }
+void gfx_draw_string_left(rct_drawpixelinfo* dpi, rct_string_id format, void* args, uint8_t colour, const ScreenCoordsXY& coords)
```
Comment (tupaschoal) [inline src/openrct2/drawing/Text.cpp:158]:
```suggestion

void gfx_draw_string_left(rct_drawpixelinfo* dpi, rct_string_id format, void* args, uint8_t colour, const ScreenCoordsXY& coords)
```
Add a line break between functions

File: src/openrct2/drawing/Drawing.h [hunk 4]
```diff
@@ -530,7 +532,7 @@ void FASTCALL
 
 // string
 void gfx_draw_string(rct_drawpixelinfo* dpi, const_utf8string buffer, uint8_t colour, int32_t x, int32_t y);
-
```
Comment (tupaschoal) [inline src/openrct2/drawing/Drawing.h:533]: Add this line back

File: src/openrct2/drawing/Text.cpp [hunk 5]
```diff
@@ -155,6 +155,10 @@ void gfx_draw_string_left(rct_drawpixelinfo* dpi, rct_string_id format, void* ar
 {
     DrawTextCompat(dpi, x, y, format, args, colour, TextAlignment::LEFT);
```
Comment (tupaschoal) [inline src/openrct2/drawing/Text.cpp:156]:
```suggestion
    gfx_draw_string_left(dpi, format, args, colour, { x, y });
```
Make the old function call the new one, so we don't have duplicate code doing the same thing and also test all of them at once. When we finish migrating to the new format, we can just delete this one and there shouldn't be any side effect.

**Comment by tupaschoal** [comment]
Part of #11571

(Another tip: if you're fixing one of our issues, the commit should read "Fix #<Issue#>: short description". If it's just one of multiple changes "Part of #<Issue#>: short description"...)

File: src/openrct2/drawing/Text.cpp (hunk 5 above)
Comment (modfort) [reply src/openrct2/drawing/Text.cpp:156]: ok thanks , i never work with clang format so it's a little bit new , for the commit i didn't understand clearly how they work from a fork sometime so i didn't really know if they are going to my fork or to this .

Comment (tupaschoal) [reply src/openrct2/drawing/Text.cpp:156]: It's just a script that applies some of our formatting rules. It checks out your fork and branch and applies the script. If there's a diff, it means it's not properly formatted, otherwise it would be blank.

Comment (modfort) [reply src/openrct2/drawing/Text.cpp:156]: Can i found a version to test it in my computer before push it .

Comment (tupaschoal) [reply src/openrct2/drawing/Text.cpp:156]: The script we run itself is over at `scripts/check-code-formatting`. You'll need to have `clang-format` installed/on your path, that varies on how to do it depending on the OS, but a few google searches will do the trick.

`clang-format` comes bundled with `LLVM`

**Comment by tupaschoal** [comment]
Please make sure that:
- format is compliant. You can check the details of the run below to see what needs to be changed
- commit message sticks to <72 chars (You might need to squash and reword your commits)
- the commit format is "Part of #NUMBER" not "#Part of NUMBER". The "#" symbol needs to come immediately before the number for GitHub to recognize it.

**Comment by modfort** [comment]
> Please make sure that:
> 
> * format is compliant. You can check the details of the run below to see what needs to be changed
ok
> * commit message sticks to <72 chars (You might need to squash and reword your commits)
ok
> * the commit format is "Part of #NUMBER" not "#Part of NUMBER". The "#" symbol needs to come immediately before the number for GitHub to recognize it.
ok
I have modify the commit but i don't know what to do for test .

File: src/openrct2-ui/windows/EditorScenarioOptions.cpp [hunk 6]
```diff
@@ -1030,49 +1030,47 @@ static void window_editor_scenario_options_guests_paint(rct_window* w, rct_drawp
...
-    x = w->windowPos.x + w->widgets[WIDX_GUEST_INITIAL_HAPPINESS].left + 1;
-    y = w->windowPos.y + w->widgets[WIDX_GUEST_INITIAL_HAPPINESS].top;
-    arg = (gGuestInitialHappiness * 100) / 255;
-    gfx_draw_string_left(dpi, STR_PERCENT_FORMAT_LABEL, &arg, COLOUR_BLACK, x, y);
-
-    // Guest initial hunger label
-    x = w->windowPos.x + 8;
-    y = w->windowPos.y + w->widgets[WIDX_GUEST_INITIAL_HUNGER].top;
-    gfx_draw_string_left(dpi, STR_GUEST_INIT_HUNGER, nullptr, COLOUR_BLACK, x, y);
+        // Guest initial happiness label
```
Comment (Gymnasiast) [inline src/openrct2-ui/windows/EditorScenarioOptions.cpp:1041]: Why did you move all these blocks into the `if (w->widgets[WIDX_CASH_PER_GUEST].type != WWT_EMPTY)` block?

File: src/openrct2/drawing/Drawing.h [hunk 7]
```diff
@@ -531,7 +532,9 @@ void FASTCALL
 // string
 void gfx_draw_string(rct_drawpixelinfo* dpi, const_utf8string buffer, uint8_t colour, int32_t x, int32_t y);
 
-void gfx_draw_string_left(rct_drawpixelinfo* dpi, rct_string_id format, void* args, uint8_t colour, int32_t x, int32_t y);
+void gfx_draw_string_left(
+    rct_drawpixelinfo* dpi, rct_string_id format, void* args, uint8_t colour, const ScreenCoordsXY& coords);
+// void gfx_draw_string_left(rct_drawpixelinfo* dpi, rct_string_id format, void* args, uint8_t colour, int32_t x, int32_t y);
```
Comment (Gymnasiast) [inline src/openrct2/drawing/Drawing.h:537]: The old definition should just be removed, no need to comment it out (we have source control to find back old code :-) ).

File: src/openrct2/drawing/Text.cpp (hunk 5 above)
Comment (Gymnasiast) [reply src/openrct2/drawing/Text.cpp:157]: @tupaschoal I'm not sure I understand. The old function with separate x and y parameters should be removed outright.

File: src/openrct2-ui/windows/Ride.cpp [hunk 8]
```diff
@@ -3881,8 +3881,9 @@ static void window_ride_operating_paint(rct_window* w, rct_drawpixelinfo* dpi)
     {
         auto blockSections = ride->num_block_brakes + ride->num_stations;
         gfx_draw_string_left(
-            dpi, STR_BLOCK_SECTIONS, &blockSections, COLOUR_BLACK, w->windowPos.x + 21,
-            ride->mode == RIDE_MODE_POWERED_LAUNCH_BLOCK_SECTIONED ? w->windowPos.y + 89 : w->windowPos.y + 61);
+            dpi, STR_BLOCK_SECTIONS, &blockSections, COLOUR_BLACK,
+            { w->windowPos.x + 21,
+              ride->mode == RIDE_MODE_POWERED_LAUNCH_BLOCK_SECTIONED ? w->windowPos.y + 89 : w->windowPos.y + 61 });
```
Comment (Gymnasiast) [inline src/openrct2-ui/windows/Ride.cpp:3886]: Could you create an intermediate assignment for the y coordinate?
`const auto z = ride->mode == RIDE_MODE_POWERED_LAUNCH_BLOCK_SECTIONED ? w->windowPos.y + 89 : w->windowPos.y + 61`

Then change this call to 
`gfx_draw_string_left(dpi, STR_BLOCK_SECTIONS, &blockSections, COLOUR_BLACK, { w->windowPos.x + 21, y });`

File: src/openrct2-ui/windows/EditorObjectiveOptions.cpp [hunk 9]
```diff
@@ -35,7 +35,8 @@
 #pragma region Widgets
 
 // clang-format off
-enum {
+enum
+{
```
Comment (Gymnasiast) [inline src/openrct2-ui/windows/EditorObjectiveOptions.cpp:39]: This file contains a ton of formatting changes that are unrelated. Please revert these.

File: src/openrct2/drawing/Text.cpp (hunk 5 above)
Comment (tupaschoal) [reply src/openrct2/drawing/Text.cpp:157]: @Gymnasiast it's easier to introduce an overload to make the migration iteratively, otherwise we have to update all calls at once. If those are a few (and already updated here), then I agree with you, but by the time I checked I think they weren't fully migrated.

In any case, if the old function is always made to call the new one, then cleaning it up later should we ever forget to would be easy.

Comment (Gymnasiast) [reply src/openrct2/drawing/Text.cpp:157]: Considering they commented it out, I'll have to assume that they migrated all calls anyway.

**Comment by tupaschoal** [comment]
Hello @modfort ,

Can you please rebase this to solve the merge conflicts and also deal with the issues?

Thanks

**Comment by modfort** [comment]
hello @tupaschoal  , 
Sorry if i did nt continue , i have some trouble this week , but whatever i check the new thing  i have to do , but i did'nt understand everything , 
"while (track_block_get_previous({ returnX, returnY, returnTrackElement }, &trackBeginEnd))" 
What i have to change with this ? My work is relative to gfx_draw_string_left? 
And a another thing , 
instead of having 
auto x = something1,
auto y = something 2
i have  to do this 
ScreenCoordsXY coords = {x , y }
 is that what i have to do ?
ps :sorry if i can appear rude, just i don't master english well enough

**Comment by tupaschoal** [comment]
> "while (track_block_get_previous({ returnX, returnY, returnTrackElement }, &trackBeginEnd))"
What i have to change with this ? My work is relative to gfx_draw_string_left?

I didn't understand if you had a problem rebasing your pull request or if this is regarding a comment I said. Because the code has been modified since you started your work, it is not cleanly applying anymore, so you need to rebase It. There will be rebase conflicts, hence it not being possible to apply cleanly, which might look like pieces of code you didn't touch.

> i have to do this
ScreenCoordsXY coords = {x , y }
is that what i have to do ?

That is preferred, as it clearly shows the meaning. You can just do it in the function parameter directly too, like:
`gfx_draw_string_left(..., { x, y }, ...)`

**Comment by janisozaur** [comment]
There are currently 55 commits in this PR, with a great lot of them not even compiling, unrelated and misnamed. The PR lists 13 conflits and is not going to be considered until these are resolved. @modfort can I please ask you at least try compiling the code before you submit it?

**Comment by modfort** [comment]
> There are currently 55 commits in this PR, with a great lot of them not even compiling, unrelated and misnamed. The PR lists 13 conflits and is not going to be considered until these are resolved. @modfort can I please ask you at least try compiling the code before you submit it?

Yes you are right i'm trying to solve the problem and trying to pushing ( i just merge the code  so i did'nt compile the code think that it's ok ) , but now i will delete this and fork another time and do it from the begining , i think it can be better ? @janisozaur

**Comment by janisozaur** [comment]
> ( i just merge the code so i did'nt compile the code think that it's ok )

Why don't you do at least a smoke test if it compiles after a merge? Doesn't make a lot of sense to send code that is guaranteed to fail on all the jobs.

> now i will delete this and fork another time and do it from the begining

Please don't.

If you wish to stop working on it, close the PR and leave a comment, but don't delete the fork. It will only create more mess.

**Comment by modfort** [comment]
> If you wish to stop working on it, close the PR and leave a comment, but don't delete the fork. It will only create more mess.
So after trying to finish the work , i don't really know how i can deal with it so i will not continue and close the pr
